### PR TITLE
Cache dataset_id lookups for logs and remove related lookups

### PR DESCRIPTION
**Pull request by ralphiee22**
### Summary

The main motivation behind this PR is to optimize deserialization (serialized store models being loaded into the django app layer). When we deserialize, for each model we call `save`, which calls `ensure_dataset`, which does lots of db related lookups. This adds up and makes this operation slow.

Performance Improvement for same number of records:
**BEFORE:**
```
('func: ', 'deserialization')
('queries:', 2955)
('READS: ', 2363)
('WRITES: ', 582)
took: 9.25s
```

**AFTER:**
```
('func: ', 'deserialization')
('queries:', 1655)
('READS: ', 1063)
('WRITES: ', 582)
took: 5.29s
```

Proposed solution in this PR:
Change related lookup calls in `infer_dataset` calls from `self.user.dataset` -> `self.user.dataset_id` (reduces read queries by one).
Also caching dataset lookup calls for logs (saves one read query per log). 

Possible point of contention: One thing is that sqlite does not enforce foreign key constraints so it may be possible to save an object with a non existent `dataset_id`.
Possible solution: We can enable foreign key constraints (if the sqlite version supports it) which would avoid the above https://www.sqlite.org/foreignkeys.html#fk_enable

### Reviewer guidance

Do tests pass?
Does using the app as normal work or noticing any strange behavior?

----

### Contributor Checklist

- [x] Contributor has fully tested the PR manually
- [x] PR has the correct target branch and milestone
- [x] PR has 'needs review' or 'work-in-progress' label
- [x] If PR is ready for review, a reviewer has been added. (Don't use 'Assignees')
- [ ] If there are any front-end changes, before/after screenshots are included
- [ ] If this is an important user-facing change, PR or related issue has a 'changelog' label

### Reviewer Checklist

- [ ] Automated test coverage is satisfactory
- [ ] Reviewer has fully tested the PR manually
- [ ] PR has been tested for [accessibility regressions](http://kolibri-dev.readthedocs.io/en/develop/manual_testing.html#accessibility-a11y-testing)
- [ ] External dependencies files were updated (`yarn` and `pip`)
- [ ] Documentation is updated
- [ ] Link to diff of internal dependency change is included
- [ ] CHANGELOG.rst is updated for high-level changes
- [ ] Contributor is in AUTHORS.rst

### CODE REVIEW

File: kolibri/core/auth/models.py [hunk 1]
```diff
@@ -159,6 +160,23 @@ class AbstractFacilityDataModel(FacilityDataSyncableModel):
...
+    def cache_related_dataset_lookup(self, related_obj_name):
+        """
+        Attempt to get the dataset_id either from the cache or the actual related obj instance.
+
+        :param related_obj_name: string representing the name of the related object on this model
+        :return: the dataset_id associated with the related obj
+        """
+        field = self._meta.get_field(related_obj_name)
+        key = '{id}_{db_table}_dataset'.format(id=getattr(self, field.attname), db_table=field.related_model._meta.db_table)
+        cached_value = cache.get(key)
```
Comment (rtibbles) [inline kolibri/core/auth/models.py]: Mostly stylistic, but I quite like doing this as a single if statement:

```suggestion
        dataset_id = cache.get(key)
        if dataset_id is None:
            dataset_id = getattr(self, related_obj_name).dataset_id
            cache.set(key, dataset_id, 60 * 10)
        return dataset_id
```

File: kolibri/core/auth/models.py [hunk 2]
```diff
@@ -159,6 +160,23 @@ class AbstractFacilityDataModel(FacilityDataSyncableModel):
     class Meta:
         abstract = True
 
+    def cache_related_dataset_lookup(self, related_obj_name):
```
Comment (rtibbles) [inline kolibri/core/auth/models.py]: Maybe `cached_related_dataset_lookup`?

**Review by rtibbles — COMMENTED** [review]
Makes sense, and the reduced reads speak for themselves, a couple of stylistic comments mostly.

**Comment by indirectlylit** [comment]
seems like a significant improvement. What needs to be done to get this merged in?

**Comment by ralphiee22** [comment]
@indirectlylit Nothing, can be merged as is.

**Review by indirectlylit — APPROVED** [review]

**Pull request merged**